internal/topology/metrics.go

```
@@ -0,0 +1,37 @@
+package topology
+
+// TODO this is only here to avoid recursive imports. We should have a separate package for metrics.
```

This was sadly needed to avoid circular imports.

---

as a curiosity, which variable requires something from topology in this file? A did a quick look but didn't find it

---

Moreover, for a cleaner decoupling btw topology & metrics, you should provide helper functions to add new data point in prometheus.
For example, in the heartbeat() function https://github.com/ContentSquare/chproxy/pull/356/files#diff-b96f6b47761d0b9abcc559863275c3c14a865426cc0ed869fee5411eb497311cR146, the function has to know that the possible labels are "cluster", "replica", "cluster_node". As a result, if you change the name of a label (or remove it) in metrics.go and forget to change it in node.go, it won't work (I don't know the behavior of prometheus.GaugeOpts but I guess it should throw an error).